backend/tracim_backend/lib/core/user.py

```diff
@@ -695,34 +706,37 @@ def saml_authenticate(
         or UserAuthenticatedIsNotActive
         :param user to check, can be none if user not found, will try
          to create new user if none but ldap auth succeed
-        :param user_id: user_id of the user
-        :param name extracted from saml attributes
+        :param external_id
+        :param username of the user
+        :param public_name extracted from saml attributes
         :param mail extracted from saml attributes
         :param profile: user Profile extracted from saml attributes
         """
         auth_type = AuthType.SAML
 
-        # TODO - M.L - 2023/09/06 - Change this to get saml response and isolate attributes
-
         if not user:
             try:
-                user = self.get_one_by_login(user_id)
+                user = self.get_one_by_external_id(external_id)
+                self.update(
+                    user=user, email=mail, username=username, name=public_name, do_save=False
+                )
             except UserDoesNotExist:
                 _ = self.create_user(
+                    external_id=external_id,
                     email=mail,
-                    username=user_id,
-                    name=name,
+                    username=username,
+                    name=public_name,
                     auth_type=AuthType.SAML,
                     do_save=True,
                     do_notify=False,
                     profile=profile,
                 )
-                transaction.commit()
-                user = self.get_one_by_login(user_id)
+                # transaction.commit()
+                user = self.get_one_by_external_id(external_id)
         if user and user.auth_type not in [auth_type, AuthType.UNKNOWN]:
             raise WrongAuthTypeForUser(
                 'User "{}" auth_type is {} not {}'.format(
-                    user_id, user.auth_type.value, auth_type.value
+                    username, user.auth_type.value, auth_type.value
```

Comment:
Username is not the right discriminant foir this case:

- if you are in the saml context, then the external_id is the right discriminant. In this case, the error must explain that the check is done on external id. Something like this: `'User with external_id "{}" auth_type is {} not {}'`
- if you are in the global login context, the tracim user_id (int) is the right discriminant

Reply:
Fixed in last commit